### PR TITLE
Merge neoden4 specific changes from MatSpy neoden4camera_jna fork 

### PR DESCRIPTION
**Pull request by jrobgh**
# Description
From [neoden4camera_jna](https://github.com/MatSpy/openpnp/tree/neoden4camera_jna) branch I picked the Neoden4 related changes from the [modifications there](https://github.com/openpnp/openpnp/compare/test...MatSpy:neoden4camera_jna). This included the machine/neoden4 folder and some files outside of that. There are a number of other modifications in that branch I did not include here such as: multi-fiducial related, precise size related, panelization, debugging logger, mousewheel zoom, etc. those can be submitted separately as they are not ND4 specific.

***Still need to figure out about how to package the NeodenCamera.dll.. happy to discuss that and the rest of the changes further, adapt as needed, etc.***

# Justification
This would enable the Neoden4 machines to be used with OpenPnP, providing ND4 users a better alternative to the stock ND software and to help grow the OpenPnP community. 

# Instructions for Use
There's a wiki page work in progress for Neoden4 users getting started [here](https://github.com/openpnp/openpnp/wiki/Neoden4-Getting-Started)

# Implementation Details
1. I tested this on my ND4 running Windows7 x32, java -version 1.8.0_121 built with Eclipse 4.9.0 (Build id: 20180917-1800). 
Connected to the machine, homed, actuated feeder.

2. Did you follow the [coding style](https://github.com/openpnp/openpnp/wiki/Developers-Guide#coding-style)?

Yes

3. If you made changes in the `org.openpnp.spi` or `org.openpnp.model` packages you will need to add additional justification for these changes. Changes to these packages require extensive review and testing.

openpnp.model.part.java modified to allow the feeder to track the part rotation and pitch in tape

4. Be sure to run `mvn test` before submitting the Pull Request. If the tests do not pass the Pull Request will not be accepted.

Passed the tests

### CODE REVIEW

File: src/main/java/org/openpnp/spi/base/AbstractPnpJobProcessor.java [hunk 1]
```diff
@@ -27,7 +27,7 @@ public static void discardAll(Head head) throws JobProcessorException {
      */
     public static void discard(Nozzle nozzle) throws JobProcessorException {
         try {
-            Cycles.discard(nozzle);
+            Cycles.discard(nozzle);
```
[the remark this answers is not in the document]
Comment (jrobgh) [reply src/main/java/org/openpnp/spi/base/AbstractPnpJobProcessor.java:30]: resolved in 7e88bda50bb7eee062702dfd26c98f883ac1195a

File: src/main/java/org/openpnp/model/Part.java [hunk 2]
```diff
@@ -49,8 +49,14 @@
...
     private int pickRetryCount = 0;
+    
+    // Part in tape attributes
+	@Attribute(required = false)
+    private int rotationInTape = 0;
 
-
+	@Attribute(required = false)
+    private int pitchInTape = 4;
+
```
[the remark this answers is not in the document]
Comment (jrobgh) [reply src/main/java/org/openpnp/model/Part.java]: Fixed in e2e84c3882637b0c028f30089b49aecf1d924e5a

File: src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java [hunk 3]
```diff
@@ -248,6 +248,30 @@ public void setHeadOffsets(Location headOffsets) {
         firePropertyChange("headOffsets", oldValue, headOffsets);
         adjustHeadOffsetsDependencies(oldValue, headOffsets);
     }
+
```
[the remark this answers is not in the document]
Comment (jrobgh) [reply src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java:251]: 40dd7725a5071b32e5c5a5f5415ced9b90dfd807

File: src/main/java/org/openpnp/machine/reference/driver/SerialPortCommunications.java [hunk 4]
```diff
@@ -137,8 +137,25 @@ public synchronized void disconnect() throws Exception {
     }
 
     public int read() throws TimeoutException, IOException {
-        byte[] b = new byte[1];
```
[the remark this answers is not in the document]
Comment (jrobgh) [reply src/main/java/org/openpnp/machine/reference/driver/SerialPortCommunications.java:140]: cc @MatSpy looks like this change is from eb199d5e48a7905eea04cda748afb598b30233c2 

This is using jSerialComm.SerialPort bytesAvailable to check if there's new serial data (non-blocking) and then sleep the thread if not.. I suspect this was done to improve multi-threaded performance on the slower ND4 Atom CPU.

It looks like the main test branch is currently using setComPortTimeouts with TIMEOUT_READ_SEMI_BLOCKING, 500 ms read timeout which would cause this thread to block up to 500ms if no data is ready... although having that non-zero, longer (500ms) timeout might improve average latency to service incoming bytes (than yielding cycles to other threads)?

I haven't tested with/without this change personally. It could potentially impact timing behavior of other machines, so I'll go ahead and move it to a separate PR to make it cleaner to test against & report back on how removing this change impacts ND4.

**Comment by jrobgh** [comment]
Thanks for the review and comments, just re-ran the maven tests and passed

**Comment by jrobgh** [comment]
Been testing this a bit without the change to the serial uart read.. I do notice some occasional pauses after part picks (not sure if this is vacuum sense related?) but haven't found the root cause yet (there is some lag in reading/writing to the vacuum actuators from the machine controls panel so might be related but haven't A/B compared yet without the serial uart read change).

It also seems that having both the Feeder Pick Location Rotation and 'Rotation in Tape' is a bit redundant.. only use case I can find is for the feeders on the right side of the machine to have 0 rotation, then left side 180? Alternately these could just be combined into the pick rotation I think which feels simpler.

File: src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (markmaker) [reply src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java:251]: You still need to remove these "Name" setter/getters here.

File: src/main/java/org/openpnp/model/Part.java [hunk 5]
```diff
@@ -50,7 +50,6 @@
     @Attribute(required = false)
     private int pickRetryCount = 0;
 
-
```
Comment (markmaker) [inline src/main/java/org/openpnp/model/Part.java:53]: Artifact.

**Review by markmaker — COMMENTED** [review]
This seems almost finished now to me.

**Comment by markmaker** [comment]
> Been testing this a bit without the change to the serial uart read.. I do notice some occasional pauses after part picks (not sure if this is vacuum sense related?) but haven't found the root cause yet (there is some lag in reading/writing to the vacuum actuators from the machine controls panel so might be related but haven't A/B compared yet without the serial uart read change).

I guess you'd have to try, but frankly I don't see how this change might help.

> ... only use case I can find is for the feeders on the right side of the machine to have 0 rotation, then left side 180? Alternately these could just be combined into the pick rotation I think which feels simpler.

Yes, this is exactly what is supposed to happen. There are three things to distinguish:

1. Feeder orientation stored in its `location.rotation`. If you rotate a feeder, it should _not_ affect the Rotation in Tape. Note, it can be a very slight rotation, e.g. with the fiducial-calibrated BlindsFeeder, or when following along bent ReferenceStripFeeder. Or a large rotation, when you mount a (slot) feeder on another machine edge.
2. Rotation in Tape is really just that, _regardless_ of feeder orientation. It is good to be separate, otherwise it would be very cumbersome the use and/or program. 
3. Pick location, is the _combination_ of the two. It is usually calculated on the fly.

_Mark

File: src/main/java/org/openpnp/model/Part.java (hunk 5 above)
Comment (jrobgh) [reply src/main/java/org/openpnp/model/Part.java]: Fixed in b7dd9a39d9c27824bea9288f75a8d8aaeb0ca33d

File: src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jrobgh) [reply src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java]: Thanks- I pulled the latest from openpnp/test, does that resolve it?

File: src/main/java/org/openpnp/machine/neoden4/wizards/Neoden4CameraConfigurationWizard.java [hunk 6]
```diff
@@ -46,8 +52,8 @@ public Neoden4CameraConfigurationWizard(Neoden4Camera camera) {
 
         panelGeneral = new JPanel();
         contentPanel.add(panelGeneral);
-        panelGeneral.setBorder(new TitledBorder(null,
-                "General", TitledBorder.LEADING, TitledBorder.TOP, null));
+        panelGeneral.setBorder(new TitledBorder(new EtchedBorder(EtchedBorder.LOWERED, null, null),
```
Comment (markmaker) [inline src/main/java/org/openpnp/machine/neoden4/wizards/Neoden4CameraConfigurationWizard.java:55]: Sorry, I missed these so far. You should search all your sources for `EtchedBorder `and get rid of them.

See #1289.

File: src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java [hunk 7]
```diff
@@ -1365,8 +1365,14 @@ public void findIssues(Solutions solutions) {
...
-                findActuatorIssues(solutions, getVacuumActuator(), "vacuum valve", 
-                        new CommandType[] { CommandType.ACTUATE_BOOLEAN_COMMAND });
+                if (getVacuumActuator().getValueType() == ActuatorValueType.Double) {
+                    findActuatorIssues(solutions, getVacuumActuator(), "vacuum valve", 
+                        new CommandType[] { CommandType.ACTUATE_DOUBLE_COMMAND });
+                }
+                else {
+                    findActuatorIssues(solutions, getVacuumActuator(), "vacuum valve", 
+                            new CommandType[] { CommandType.ACTUATE_BOOLEAN_COMMAND });
+                }
```
Comment (markmaker) [inline src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java:1375]: Do you know, why these changes pop up here? They are already in the `test` branch. 

Whitespace? Line-endings?

**Comment by MatSpy** [comment]
> Yes, this is exactly what is supposed to happen. There are three things to distinguish:
> 
>     1. Feeder orientation stored in its `location.rotation`. If you rotate a feeder, it should _not_ affect the Rotation in Tape. Note, it can be a very slight rotation, e.g. with the fiducial-calibrated BlindsFeeder, or when following along bent ReferenceStripFeeder. Or a large rotation, when you mount a (slot) feeder on another machine edge.
> 
>     2. Rotation in Tape is really just that, _regardless_ of feeder orientation. It is good to be separate, otherwise it would be very cumbersome the use and/or program.
> 
>     3. Pick location, is the _combination_ of the two. It is usually calculated on the fly.

Hello Mark,

Yes exactly this is correct approach. It maters a lot when you want to rerun project and/or you have multiple project with same parts. You put simply tape to any feeder and pick orientation is calculated automatically. You have to once enter correct orientation on tape at start.

File: src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java (hunk 7 above)
Comment (jrobgh) [reply src/main/java/org/openpnp/machine/reference/ReferenceNozzle.java]: thanks- I needed to merge in the latest openpnp:test

**Review by markmaker — APPROVED** [review]

**Comment by markmaker** [comment]
Please update the [Wiki](https://github.com/openpnp/openpnp/wiki/Neoden4-Getting-Started), so users know they can now just download the testing version. Personally, I think you guys should really invest some time in the Wiki, add screenshots and photos/videos of the machines! Make some announcements on the mailing list, so all the new good code is properly advertised!

_Mark
